lib/Dialect/DC/DCOps.cpp

```diff
@@ -0,0 +1,195 @@
+  // Unpack of a pack is a no-op.
+  if (auto pack = getInput().getDefiningOp<PackOp>()) {
+    results.push_back(pack.getToken());
+    results.append(pack.getInputs().begin(), pack.getInputs().end());
+    return success();
+  }
+
+  return failure();
+}
+
+LogicalResult UnpackOp::inferReturnTypes(
+    MLIRContext *context, std::optional<Location> loc, ValueRange operands,
+    DictionaryAttr attrs, mlir::OpaqueProperties properties,
+    mlir::RegionRange regions, SmallVectorImpl<Type> &results) {
+  auto inputType = operands.front().getType().cast<ValueType>();
+  results.push_back(TokenType::get(context));
+  results.append(inputType.getInnerTypes().begin(),
+                 inputType.getInnerTypes().end());
+  return success();
+}
+
+// =============================================================================
+// PackOp
+// =============================================================================
+
+void PackOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                         MLIRContext *context) {}
+
+OpFoldResult PackOp::fold(FoldAdaptor adaptor) {
+  auto token = getToken();
+  auto inputs = getInputs();
+
+  // Pack of an unpack is a no-op.
+  if (auto unpack = token.getDefiningOp<UnpackOp>()) {
+    llvm::SmallVector<Value> unpackResults = unpack.getResults();
+    if (unpackResults.size() == inputs.size() &&
+        llvm::all_of(llvm::zip(getInputs(), unpackResults), [&](auto it) {
+          return std::get<0>(it) == std::get<1>(it);
+        })) {
+      return unpack.getInput();
+    }
+  }
+
+  return {};
+}
+
+LogicalResult PackOp::inferReturnTypes(
+    MLIRContext *context, std::optional<Location> loc, ValueRange operands,
+    DictionaryAttr attrs, mlir::OpaqueProperties properties,
+    mlir::RegionRange regions, SmallVectorImpl<Type> &results) {
+  llvm::SmallVector<Type> inputTypes;
+  for (auto t : operands.drop_front().getTypes())
+    inputTypes.push_back(t);
+  auto valueType = dc::ValueType::get(context, inputTypes);
+  results.push_back(valueType);
+  return success();
+}
+
+// =============================================================================
+// SelectOp
+// =============================================================================
+
+void SelectOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                           MLIRContext *context) {}
+
+// =============================================================================
+// BufferOp
+// =============================================================================
+
+FailureOr<SmallVector<int64_t>> BufferOp::getInitValueArray() {
+  assert(getInitValues() && "initValues attribute not set");
+  SmallVector<int64_t> values;
+  for (auto value : getInitValuesAttr()) {
+    if (auto iValue = value.dyn_cast<IntegerAttr>()) {
+      values.push_back(iValue.getValue().getSExtValue());
+    } else {
+      return emitError() << "initValues attribute must be an array of integers";
```

Comment:
Better, but in what case is this assumption reasonable? I suspect that outside of testing, this case would be rare.